Feature/sirius context and config

aminaramoon opened the pull request:
Features addded:

- added a confiration manager using libconfig that allows configuration to be set from different libraries without injecting dependency to the third-party libraries : i.e. you can configure cucascade without adding the configuration manager to the cucascade.
- added sirius_context, a global singleton store for all engine resource
- added sirius_context_lifetime_manager that manages lifetime of the states, begin/end queries, configuring the engine using a config file.

joosthooz commented on src/sirius_context.cpp:
```diff
@@ -0,0 +1,226 @@
+  return *task_creator_;
+}
+
+void SiriusContext::thorw_if_not_initialized() const
```
typo - throw

joosthooz commented on src/include/pipeline/config.hpp:
```diff
@@ -0,0 +1,39 @@
+    pipeline_task_executor_config.num_threads    = 4;
+    pipeline_task_executor_config.retry_on_error = false;
+
+    gpu_task_executor_config.num_threads    = 2;
```
What is the difference between the pipeline and gpu task executors? Why does one have 2 threads and the other 4?

aminaramoon committed:
- added sirius context, starting up all the components
- injecting deps to all high level componenets
- creating locks for extension, so only one connection can be loaded
- created cudf memory_resource that checks the device and uses correct allocators.
- added config for sirius, engine can be loaded in config file is specified

aminaramoon committed:
add github action to use pixi

aminaramoon committed:
push updates cucascade to sirius

aminaramoon committed:
disable old test

mbrobbel committed:
Merge branch 'dev' into feature/sirius_context_and_config

mbrobbel committed:
Use cuda12 environment for test workflow

mbrobbel committed:
Overwrite unzip for dbgen in test workflow

mbrobbel committed:
Use setup_test_datasets script in test workflow

joosthooz commented on src/creator/task_creator.cpp:
```diff
@@ -100,23 +114,23 @@ void task_creator::process_next_task(op::sirius_physical_operator* node)
   } else {
-    if (!priority_scans.empty()) {
-      duckdb::shared_ptr<pipeline::sirius_pipeline> pipeline = priority_scans.front();
+    if (!_priority_scans.empty()) {
```
Not about this PR specifically, but if we are doing `while (!_priority_scans.empty()) {` here https://github.com/sirius-db/sirius/pull/133/changes#diff-daca317e191e57036793b6c21fb12b69ca29699cfe5609bf34b8bf78f3cd54a3R129, shouldn't `_priority_scans` be unneeded here? Or will there be additional tasks pushed into it?

joosthooz replied:
Also, if this function is to be called by an executor thread, we need to make these queue accesses thread-safe
